### PR TITLE
Group colours by regex of the experiment name.

### PR DESCRIPTION
**Pull request by AnuarTB**
## Motivation for features / changes

Users want to color runs by the regex string of the corresponding to runs experiment names. 

## Technical description of changes

- Added new GroupBy type, REGEX_BY_EXP.
- Added a dropdown in dialog window, so users could select between regex for run name or experiment name. 

## Screenshots of UI changes (or N/A)

N/A since internal change. 

## Detailed steps to verify changes work correctly (as executed by you)

- Run tensorboard.corp server.
- Click on color grouping icon and select `Regex`.
- Select `Experiment Name` in dropdown.

## Alternate designs / implementations considered (or N/A)
N/A

### CODE REVIEW

**Review by hoonji — CHANGES_REQUESTED** [review]
This PR does not follow our review best practices. Can you please first 1) remove (or split out) the unnecessary changes like BUILD file reordering and 2) at least split the made code into two separate PRs (e.g. redux code first, then the rest) (go/small-cls)

**Comment by AnuarTB** [comment]
@hoonji I will certainly try then remove the formatting of the BUILD files in this PR. Regarding the change you requested. 

>  2) at least split the made code into two separate PRs (e.g. redux code first, then the rest)

Does it just mean separate the formatting of the files and added feature code?

**Comment by hoonji** [comment]
@AnuarTB I mean separating out the actions, effects, reducers and utils into a separate PR, and putting the component/container logic in a followup PR.

I think it's silly to spend too much effort just splitting code, but we should prefer it in cases like this where a split is easy and natural.

**Comment by roseayeon** [comment]
> I mean separating out the actions, effects, reducers and utils into a separate PR, and putting the component/container logic in a followup PR.


+1 This PR is quite large.  It would be very helpful to review if you separated this PR into two parts.

**Comment by AnuarTB** [comment]
Ok so If I understood correctly, I should separate store, actions and reducers into first PR and then the views in the second PR? If that's so I will separate now.

**Comment by roseayeon** [comment]
> I should separate store, actions and reducers into first PR and then the views in the second PR?

Yeah Exactly 

 > If that's so I will separate now, I was just confused on how should i separate this CL, as it is one big coherent block of code.
 How about putting all the other code in one PR, and the code under `views/runs_table` in a second PR? Does it sound okay?

File: tensorboard/webapp/runs/store/runs_reducers_test.ts [hunk 1]
```diff
@@ -158,6 +158,7 @@ describe('runs_reducers', () => {
             ],
           },
         },
+        expNameByExpId: {},
```
Comment (roseayeon) [inline tensorboard/webapp/runs/store/runs_reducers_test.ts:161]: QQ. If it's optional, don't we need to add this line?

**Review by roseayeon — COMMENTED** [review]
Congratulations on adding a new feature to TB!

File: tensorboard/webapp/runs/store/runs_reducers_test.ts (hunk 1 above)
Comment (AnuarTB) [reply tensorboard/webapp/runs/store/runs_reducers_test.ts:161]: Oh, I have question. Is it an ok idea to make a field in the action, to be optional? I declared it as optional, but I am not sure whether it is an ok practice. wdyt?

File: tensorboard/webapp/angular/BUILD [hunk 2]
```diff
@@ -132,6 +132,15 @@ tf_ts_library(
     ],
 )
 
+# This is a dummy rule used as a @angular/material/chips dependency.
+tf_ts_library(
+    name = "expect_angular_material_core",
```
[the remark this answers is not in the document]
Comment (AnuarTB) [reply tensorboard/webapp/angular/BUILD]: Have moved it to next PR.

File: tensorboard/webapp/runs/effects/runs_effects.ts [hunk 3]
```diff
@@ -217,12 +218,18 @@ export class RunsEffects {
         }
         return {newRuns, runsForAllExperiments};
       }),
-      tap(({newRuns, runsForAllExperiments}) => {
+      withLatestFrom(this.store.select(getDashboardExperimentNames)),
+      map(([runsData, expNameByExpId]) => {
```
Comment (hoonji) [inline tensorboard/webapp/runs/effects/runs_effects.ts:222]: This map() essentially just renames variables. Can we just include this logic in the tap() below?

Comment (hoonji) [inline tensorboard/webapp/runs/effects/runs_effects.ts:222]: nit: this map() essentially just renames variables. Can we just include this logic in the tap() below?

File: tensorboard/webapp/runs/store/runs_reducers_test.ts (hunk 1 above)
Comment (hoonji) [reply tensorboard/webapp/runs/store/runs_reducers_test.ts:161]: Optional action props should be fine conceptually and are [already used elsewhere](https://github.com/tensorflow/tensorboard/blob/d4c5c825a5677bf1d22d1d9290bb439f0ce7e861/tensorboard/webapp/runs/actions/runs_actions.ts#L90)

**Review by hoonji — APPROVED** [review]
Looks solid to me. Thanks for applying this!

File: tensorboard/webapp/runs/effects/runs_effects.ts (hunk 3 above)
Comment (roseayeon) [reply tensorboard/webapp/runs/effects/runs_effects.ts:222]: +1

**Review by roseayeon — APPROVED** [review]
LGTM!

File: tensorboard/webapp/runs/store/runs_reducers_test.ts (hunk 1 above)
Comment (AnuarTB) [reply tensorboard/webapp/runs/store/runs_reducers_test.ts:161]: Ok, removed empty dictionaries.

File: tensorboard/webapp/runs/effects/runs_effects.ts (hunk 3 above)
Comment (AnuarTB) [reply tensorboard/webapp/runs/effects/runs_effects.ts:222]: Done.

**Pull request merged**